### ISSUE TITLE
e2e is flakey with spec.clusterIP value failures

### ISSUE THREAD

**Issue by krousey** [GoogleCloudPlatform/kubernetes#9444]
Errors similar to this have been showing up in E2E runs lately. First noticed on June 8th, 2015. A sample follows from  kubernetes-e2e-gce build 6780:

`Service "nettest" is invalid: spec.clusterIP: invalid value '': cannot allocate resources of type serviceipallocation at this time`

**Comment by krousey** [comment]
cc @quinton-hoole

**Comment by krousey** [comment]
Likely causes (from scrubbing test logs and correlating with merges) are #9392 or #9390. We've disabled the e2e test in the parallel suite and still see this error, so it is likely #9390. Seems the message comes from here: https://github.com/GoogleCloudPlatform/kubernetes/blob/51750a4680cbb161b068b7748c8d00cacb8a893a/pkg/registry/service/allocator/etcd/etcd.go#L147

cc @lavalamp @fgrzadkowski

**Comment by lavalamp** [comment]
I think if this is related to @fgrzadkowski's change, it's just exposing preexisting bugs.

**Comment by fgrzadkowski** [comment]
I think this was solved by reverting my PR. Can we close it? @krousey

**Comment by krousey** [comment]
This particular issue is done, but if we plan to try that PR again, it'd be nice to know why this happened.

**Comment by smarterclayton** [comment]
We're running with the ugorji code in etcd and haven't hit this yet.  Will try to stress test it

**Comment by smarterclayton** [comment]
@derekwaynecarr we need to try and expose this

**Comment by fgrzadkowski** [comment]
I was trying to reproduce this, but failed. I've run e2e tests 4 times and all passed. I suggest we remerge PR with newer etcd client version with some additional debugging. I'll merge it my morning and will watch jenkins closely to look for regressions. If needed I'll revert it again (probably even before US morning), but hopefully with more data for debugging. WDYT? @quinton-hoole @smarterclayton @saad-ali (current oncall)

**Comment by smarterclayton** [comment]
Sgtm

**Comment by quinton-hoole** [comment]
SGTM too @fgrzadkowski

**Comment by saad-ali** [comment]
@fgrzadkowski let me know when you plan to do this, so that I don't merge other PRs at the same time

**Comment by fgrzadkowski** [comment]
Finally I've been able to reproduce this problem on Jenkins (after fixing some bugs with collecting logs). In the failed runs I see the following errors:

```
E0617 21:23:46.469886       5 controller.go:78] Unable to perform initial IP allocation check: unable to refresh the service IP block: 501: All the given peers are not reachable (failed to propose on members [http://127.0.0.1:4001] twice [last error: Get http://127.0.0.1:4001/v2/keys/registry/ranges/serviceips?quorum=false&recursive=false&sorted=false: dial tcp 127.0.0.1:4001: connection refused]) [0]
E0617 21:23:46.495514       5 controller.go:81] Unable to perform initial service nodePort check: unable to refresh the port block: 501: All the given peers are not reachable (failed to propose on members [http://127.0.0.1:4001] twice [last error: Get http://127.0.0.1:4001/v2/keys/registry/ranges/servicenodeports?quorum=false&recursive=false&sorted=false: dial tcp 127.0.0.1:4001: connection refused]) [0]
E0617 21:23:46.547455       5 controller.go:84] Unable to perform initial Kubernetes service initialization: 501: All the given peers are not reachable (failed to propose on members [http://127.0.0.1:4001] twice [last error: Put http://127.0.0.1:4001/v2/keys/registry/namespaces/default?prevExist=false: dial tcp 127.0.0.1:4001: connection refused]) [0]
```

As a result this data are not written to etcd and consecutive updates fail and master cannot assign clusterIP.

@xiang90 - Is it possible this is caused by some other etcd client library changes (other than using ugorji code, which I think is unrelated). I'm trying to update from version ```v2.0.0-3-g0424b5f``` to ```v2.0.0-13-g4cceaf7```. See #9862 for details.

@smarterclayton I think we should loop over ```RunOnce()``` calls ([here](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/master/controller.go#L77)) until we actually succeed or ```panic()``` after some longer deadline. Otherwise we may end up with completely unusable apiserver. WDYT?

**Comment by smarterclayton** [comment]
----- Original Message -----
> Finally I've been able to reproduce this problem on Jenkins (after fixing
> some bugs with collecting logs). In the failed runs I see the following
> errors:
> 
> ```
> E0617 21:23:46.469886       5 controller.go:78] Unable to perform initial IP
> allocation check: unable to refresh the service IP block: 501: All the given
> peers are not reachable (failed to propose on members
> [http://127.0.0.1:4001] twice [last error: Get
> http://127.0.0.1:4001/v2/keys/registry/ranges/serviceips?quorum=false&recursive=false&sorted=false:
> dial tcp 127.0.0.1:4001: connection refused]) [0]
> E0617 21:23:46.495514       5 controller.go:81] Unable to perform initial
> service nodePort check: unable to refresh the port block: 501: All the given
> peers are not reachable (failed to propose on members
> [http://127.0.0.1:4001] twice [last error: Get
> http://127.0.0.1:4001/v2/keys/registry/ranges/servicenodeports?quorum=false&recursive=false&sorted=false:
> dial tcp 127.0.0.1:4001: connection refused]) [0]
> E0617 21:23:46.547455       5 controller.go:84] Unable to perform initial
> Kubernetes service initialization: 501: All the given peers are not
> reachable (failed to propose on members [http://127.0.0.1:4001] twice [last
> error: Put
> http://127.0.0.1:4001/v2/keys/registry/namespaces/default?prevExist=false:
> dial tcp 127.0.0.1:4001: connection refused]) [0]
> ```
> 
> As a result this data are not written to etcd and consecutive updates fail
> and master cannot assign clusterIP.
> 
> @xiang90 - Is it possible this is caused by some other etcd client library
> changes (other than using ugorji code, which I think is unrelated). I'm
> trying to update from version ```v2.0.0-3-g0424b5f``` to
> ```v2.0.0-13-g4cceaf7```. See #9862 for details.
> 
> @smarterclayton I think we should loop over ```RunOnce()``` calls
> ([here](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/master/controller.go#L77))
> until we actually succeed or ```panic()``` after some longer deadline.
> Otherwise we may end up with completely unusable apiserver. WDYT?

Agree - transient failures here should block startup of the API server.  We don't have to panic if we move this up in the code, try a few times, then return an error if we can't proceed.  That'll exit the apiserver and force a restart (which in a containerized master environment would give us a chance to pick up new env vars or connections).

**Comment by fgrzadkowski** [comment]
Fixed by #9862
